src/ejabberd_receiver.erl

```diff
@@ -421,3 +423,15 @@ gen_server_call_or_noproc(Pid, Message) ->
     end.
 
 gen_fsm() -> p1_fsm.
+
+hibernate(Result) ->
+    case process_info(self(), message_queue_len) of
```

Comment:
With p1_fsm some messages can be inside an internal message queue.
This means that you should check '$internal_queue_len' process dictionary entry too.

I would rather put hibernate logic inside p1_fsm_old.

Reply:
:+1: for moving the logic to `p1_fsm_old`.